sklearn_hierarchical_classification/classifier.py

```diff
@@ -179,16 +201,20 @@ def fit(self, X, y=None, sample_weight=None):
             for node in self.graph_.nodes()
             if node != self.root
         )
-        self.logger.debug("fit() - self.classes_ = %s", self.classes_)
 
         # Recursively build training feature sets for each node in graph
-        with self._progress(total=self.n_classes_ + 1, desc="Building features") as progress:
-            self._recursive_build_features(X, y, node_id=self.root, progress=progress)
+        if not self.preprocessing:
+            with self._progress(total=self.n_classes_ + 1, desc="Building features") as progress:
+                self._recursive_build_features(X, y, node_id=self.root, progress=progress)
 
         # Recursively train base classifiers
         with self._progress(total=self.n_classes_ + 1, desc="Training base classifiers") as progress:
             self._recursive_train_local_classifiers(X, y, node_id=self.root, progress=progress)
 
+        #TODO future feature, using the label cardinality to calculate threshold (ML
+        if len(y.shape)==2:
+            self.avglabs = np.mean(y.sum(1))
```

Comment:
@fbenites  - `avglabs` is a bit vague of a name plus it seems based on comment this is not really used anywhere yet - safe to remove these lines completely? we can capture as separate GitHub Issue a feature request to add something like this with more details around use case..

Reply:
it is intended for a threshold procedure, I did implement as a post processing but it needs to be calculated when training. It can be implemented in the predict procedure (but needs this information). I am not sure how to integrate post-processing in the predict (sklearn does not allow multiple predict, the last step must be a predict, the others can be transform). But for the moment it can be removed.  @offtopic, thanks for the library, I won the GermEval 2019 with this library (and the post-processing). Should I link the paper when it is published?

Reply:
That’s awesome to hear! Absolutely any link appreciated

Reply:
http://arxiv.org/abs/1908.06493

Reply:
code will be available later (with blob and etc.)

Reply:
@fbenites  - thanks, very cool to see and thank you for the contribution! I am planning to merge this in with some slight touch ups hopefully by end of week. One thing that would be very useful is some example code and/or unit-test to demonstrate usage of your functionality. If you have some time or any code already using this you could contribute, would be greatly appreciated.

Reply:
I will put in my todo list (I know how important that stuff is, in principle, I checked that the code is compatible with the unit tests though). When I write the blog, I will have more time for the clean up and improvements (e.g., gridsearch per node). I think in 2 weeks I will get to that, I was also hoping to have an extended tests where datasets are downloaded and checked (like 20news groups, you can easily put an hierarchy over that, or reuters (RCV1-v2)).